app/models/user.rb

```diff
@@ -351,4 +251,132 @@ def pcode3
   def pcode4
     school.sierra_code
   end
+
+
+  # Another piece of code has determined this user is all set to be called
+  # complete and done.  Note: Usually, expect the next step to be the calling of
+  # Patron Service to create the user in Sierra.
+  def save_as_complete
+    self.status = STATUS_LABELS['complete']
+    self.save
+  end
+
+
+  # If the user's barcode is not yet finalized, then set its status to
+  # 'pending' and save.  In the future, there may be other conditions that
+  # could set the user to "pending", and we'll be checking for those here, as well.
+  def save_as_pending
+    # do we need to fill in a provisional barcode?
+    unless self.barcode.present?
+      self.barcode = self.assign_barcode
+    end
+
+    self.status = STATUS_LABELS['barcode_pending']
+    self.save
+  end
+
+
+  # ################ THE BARCODES SECTION! ################
+
+  def assign_barcode
+    LogWrapper.log('DEBUG', {
+       'message' => "Begin assigning barcode to #{self.email}",
+       'method' => "assign_barcode",
+       'status' => "start",
+       'user' => {email: self.email}
+      })
+
+    last_user_barcode = User.where('barcode < 27777099999999').order(:barcode).last.barcode
+    self.assign_attributes({ barcode: last_user_barcode + 1})
+
+    LogWrapper.log('DEBUG', {
+       'message' => "Barcode has been assigned to #{self.email}",
+       'method' => "assign_barcode",
+       'status' => "end",
+       'barcode' => "#{self.barcode}",
+       'user' => {email: self.email}
+      })
+    return self.barcode
+  end
+
+
+  def barcode_found_in_sierra
+    # Getter for flag that reflects whether there's a user or users in Sierra
+    # that correspond(s) to this user object in MLN db.
+    # NOTE: The barcode_found_in_sierra is not guaranteed to accurately reflect
+    # user sync status, until after you've called check_barcode_uniqueness_with_sierra.
+    if @barcode_found_in_sierra.blank?
+      return false
+    end
+
+    return @barcode_found_in_sierra
+  end
+
+
+  def check_barcode_uniqueness_with_sierra(barcode_to_check)
+    # Ask the platform microservice api to ask Sierra if there is already a user
+    # with the passed-in barcode.
+    # Most users will have an equivalent user record in Sierra, if set up correctly.
+    # New users and users whose records have been purged from Sierra might not.
+    # Return "true" if a user is found, false otherwise.  Default to "false".
+    # Throw an exception if called with malformed data.
+
+    if barcode_to_check.blank?
+      return false
+    end
+
+    @barcode_found_in_sierra = false
+
+    response = HTTParty.get(
+      ENV['PATRON_MICROSERVICE_URL_V01'] + "?barcode=#{barcode_to_check}",
+      headers:
+        { 'Authorization' => "Bearer #{Oauth.get_oauth_token}",
+          'Content-Type' => 'application/json' },
+      timeout: 10
+    )
+
+    case response.code
```

Comment:
Really not a big deal, and there are a dozen valid ways to write this so I'm not asking for a change, but I wouldn't use a case statement here. I'd probably go with something like:

```
statuses_for_found_barcode = [200, 409]
statuses_for_available_barcode = [404]

expected_statuses = statuses_for_found_barcode + statuses_for_available_barcode

@barcode_found_in_sierra = true if statuses_for_found_barcode.include? response.code
  
LogWrapper.log('DEBUG', {
  'method' => "#{model_name}.check_barcode_uniqueness_with_sierra",
  'message' => message_for_response_code(response.code, barcode_to_check)
  'status' => response.code
})

raise Exceptions::InvalidResponse, "Invalid status code of: #{response.code}"  unless expected_statuses.include? response.code
```

A case statement could well be of use in the hypothetical `message_for_response_code` method though.

Something like that would feel a bit more concise and less repetitive on the logging lines, and the named variables would make the intent clearer to me.

As I say, just a suggestion for discussion, rather than a request to change. Please let me know what you think!

Reply:
I agree that logging lines seem repetitive, but we frequently end up needing to perform slightly different logic for each http status.  In this code, we might needing to revise, both for clarity and level of logging, and also for what we consider "found".

Reply:
Fair enough. As I said, not asking for a change. A couple of observations and one question: I think writing code for what we might need in the future is an anti-pattern of development. Secondly, for me, the most important question is "what is the important code that another engineer should see"; in this case I'd say that it's the different classes of response codes and the `@barcode_found_in_sierra` instance variable. I think in the original those get a little lost in the log messages, and so make a new engineer work harder. It's not a big deal, still not asking for any changes, but something too think about.